pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -707,73 +709,96 @@ private void completeConnect(int clientProtoVersion, String clientVersion, boole
     }
 
     // According to auth result, send newConnected or newAuthChallenge command.
-    private State doAuthentication(AuthData clientData,
-                                   int clientProtocolVersion,
-                                   String clientVersion) throws Exception {
-
+    private CompletableFuture<Void> doAuthenticationAsync(AuthData clientData, int clientProtocolVersion,
+                                                          String clientVersion) {
         // The original auth state can only be set on subsequent auth attempts (and only
         // in presence of a proxy and if the proxy is forwarding the credentials).
         // In this case, the re-validation needs to be done against the original client
-        // credentials.
-        boolean useOriginalAuthState = (originalAuthState != null);
-        AuthenticationState authState =  useOriginalAuthState ? originalAuthState : this.authState;
-        String authRole = useOriginalAuthState ? originalPrincipal : this.authRole;
-        AuthData brokerData = authState.authenticate(clientData);
-
-        if (log.isDebugEnabled()) {
-            log.debug("Authenticate using original auth state : {}, role = {}", useOriginalAuthState, authRole);
-        }
+        // credentials, but we only can new an authentication state, because some authentication data(TLS, SASL)
+        // based on outside service.
+        // If we can get the role from the authentication sate, the global variable need to be updated.
 
-        if (authState.isComplete()) {
-            // Authentication has completed. It was either:
-            // 1. the 1st time the authentication process was done, in which case we'll send
-            //    a `CommandConnected` response
-            // 2. an authentication refresh, in which case we need to refresh authenticationData
-
-            String newAuthRole = authState.getAuthRole();
-
-            // Refresh the auth data.
-            this.authenticationData = authState.getAuthDataSource();
```

Comment:
I spent many hours working on the authentication framework code today. Below are some of my thoughts. They aren't necessarily my final opinion.

In the PR description, this line was identified as the root cause of the issue, but based on the description and on this PR, I think I see it as a symptom of a larger issue. We do not have a clear enough definition for the state transitions in the `ServerCnx` class, which is essentially a finite state machine.

I want to describe part of the problem with this line of code. On the first pass through, the `this.authenticationData` is set by getting it from `this.authState.getAuthDataSource()` because `useOriginalAuthState` is `false`. Then, on subsequent `AuthResponse` commands from the original client, the `authenticationData` is set by getting it from `this.originalAuthState.getAuthDataSource()` because `useOriginalAuthState` is `true`. That means the `this.authenticationData` is incorrectly updated.

As a note, the broker gets `AuthResponse` commands from the original client when the connection through the client is in state `ProxyConnectionToBroker`. As of https://github.com/apache/pulsar/pull/17831, the broker also gets an `AuthResponse` from the original client when the proxy is in state `ProxyLookupRequests` with `forwardAuthorizationCredentials=true`.

Otherwise, the proxy sends its own authentication information.

Here are some problems with the current solution:

1. The `AuthResponse` protocol message only has one field for `AuthData`, and there is no indication whether the `AuthData` is for the proxy or the original client. As a consequence, the broker does not know whether to update the `authenticationData` or the `originalAuthData`.
2. The implementation for `getAuthDataSource` does not always get the most recent `authenticationDataSource` See `TokenAuthenticationState`. (I am already working on fixing this.)

Open questions:
1. Can we just remove the `originalAuthState` and only keep track of one `authState`?
2. Is the `AuthenticationState` object meant to last the whole lifecycle of a given `ServerCnx`? In my mind, the answer is yes, but this PR says otherwise. If not, it feels odd that we have a "state" object.

Reply:
All of your description is correct.

> 1. The `AuthResponse` protocol message only has one field for `AuthData`, and there is no indication whether the `AuthData` is for the proxy or the original client. As a consequence, the broker does not know whether to update the `authenticationData` or the `originalAuthData`.

I started PIP a few months ago, see https://github.com/apache/pulsar/issues/17517

> 2. 1. Can we just remove the `originalAuthState` and only keep track of one `authState`?

No. We must check both the `originalAuthState` and `authState`, if you don't check, you will overstep your authority.

> 2. Is the `AuthenticationState` object meant to last the whole lifecycle of a given `ServerCnx`? In my mind, the answer is yes, but this PR says otherwise. If not, it feels odd that we have a "state" object.

When authentication data changes, we need to renew the authentication state, and then we always get the correct data from that.

Reply:
> No. We must check both the `originalAuthState` and `authState`, if you don't check, you will overstep your authority.

I agree that we currently check authorization on the proxy role and the original principal as well as with the proxy's `authenticationDataSource` and the `originalAuthenticationDataSource`. However, I am not sure why that is necessary, and as we've discussed, this is not even done correctly for the auth data source.

What is the case that a client will overstep its authority?

When the proxy is forwarding authentication data, it seems sufficient to verify that the proxy's authentication data is valid for a proxy as a one time check during connection initialization. All subsequent auth challenges will go to the client anyway.

If the proxy is unable to forward authentication data, then the broker will only have the proxy's authentication data and the `originalPrincipal`.

Reply:
> However, I am not sure why that is necessary, and as we've discussed, this is not even done correctly for the auth data source.

For the current design, the `authenticationDataSource` is credible by the proxy forwarding authentication data. This authentication data was completed in the proxy and then sent this authentication data to the broker, so we don't check the authentication data in the broker.

> When the proxy is forwarding authentication data, it seems sufficient to verify that the proxy's authentication data is valid for a proxy as a one time check during connection initialization. All subsequent auth challenges will go to the client anyway.

I thought the same thing, but the Proxy's role usually is a superuser, if we don't check the client's role, this is overstepping its authority, and the client can get the lookup data and topic data. 

If we don't consider this case, I agree with you.

> No. We must check both the originalAuthState and authState, if you don't check, you will overstep your authority.

I agree that we currently check authorization on the proxy role and the original principal as well as with the proxy's authenticationDataSource and the originalAuthenticationDataSource. However, I am not sure why that is necessary, and as we've discussed, this is not even done correctly for the auth data source.

> What is the case that a client will overstep its authority?

When the proxy is forwarding authentication data, it seems sufficient to verify that the proxy's authentication data is valid for a proxy as a one time check during connection initialization. All subsequent auth challenges will go to the client anyway.

> If the proxy is unable to forward authentication data, then the broker will only have the proxy's authentication data and the `originalPrincipal`.

Right.

Reply:
> I thought the same thing, but the Proxy's role usually is a superuser, if we don't check the client's role, this is overstepping its authority, and the client can get the lookup data and topic data.
> 
> If we don't consider this case, I agree with you.

Can you describe the attack vector? From my perspective, I don't see how a client would be able to assume the proxy's role.

Reply:
> > When the proxy is forwarding authentication data, it seems sufficient to verify that the proxy's authentication data is valid for a proxy as a one time check during connection initialization. All subsequent auth challenges will go to the client anyway.
> 
> This is `ProxyConnectionToBroker` operation, we also consider `ProxyLookupRequests` logic.

After https://github.com/apache/pulsar/pull/17831, when a proxy connection is in state `ProxyLookupRequests` with `forwardAuthorizationCredentials=true`, the proxy connection responds to the `AuthChallenge` with the original client's auth data by forwarding the broker's challenge to the client. That is why I said the auth challenges go to the client in that case.

> Would you like to pass only the proxy's authentication data? If yes, that's going to cross the line.

No, that's not my suggestion. I am suggesting that when the proxy sends both its `authData` and the client's `authData`, the broker needs to only verify the proxy's `authData` at the beginning and at no other time because in all of those cases, the broker's auth challenge is going back to the original client and the proxy's `authData` is never again updated. This is the case because the proxy only forwards the client's auth data when `forwardAuthorizationCredentials=true`. One issue with my proposed solution is that the solution introduced in https://github.com/apache/pulsar/pull/17831 is not present in older versions of the proxy, so we'd need a way to address those proxies.

Reply:
Good explanation!